File: Leetcode-one-by-one/src/Minimized Maximum of Products Distributed to Any Store.java
// https://leetcode.com/problems/minimized-maximum-of-products-distributed-to-any-store/description/

class Solution {
    public int minimizedMaximum(int n, int[] quantities) {
        int left = 1;
        int right = 0;

        for (int q : quantities) {
            right = Math.max(right, q);
        }

        while (left < right) {
            int mid = left + (right - left) / 2;

            long storesNeeded = 0;
            for (int q : quantities) {
                storesNeeded += (q + mid - 1) / mid;
            }

            if (storesNeeded <= n) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
}

File: Leetcode-one-by-one/src/Minimized Maximum of Products Distributed to Any Store.go
func minimizedMaximum(n int, quantities []int) int {
	left, right := 1, 0

	for _, q := range quantities {
		if q > right {
			right = q
		}
	}

	for left < right {
		mid := left + (right-left)/2

		var storesNeeded int64 = 0
		for _, q := range quantities {
			storesNeeded += int64((q + mid - 1) / mid)
		}

		if storesNeeded <= int64(n) {
			right = mid
		} else {
			left = mid + 1
		}
	}

	return left
}
